// react-native-calendar-range-picker/dist/Day.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React, { memo } from "react";
import { View, Text } from "react-native";
import moment from "moment";
function Day(_a) {
    var day = _a.day, locale = _a.locale, disabledBeforeToday = _a.disabledBeforeToday, style = _a.style;
    var date = day.date, type = day.type, isHoliday = day.isHoliday, isToday = day.isToday, isBeforeToday = day.isBeforeToday;
    var dayTextColor = (style === null || style === void 0 ? void 0 : style.dayTextColor) || "#1d1c1d";
    var holidayColor = (style === null || style === void 0 ? void 0 : style.holidayColor) || "#f26522";
    var todayColor = (style === null || style === void 0 ? void 0 : style.todayColor) || "#1692e4";
    var selectedDayTextColor = (style === null || style === void 0 ? void 0 : style.selectedDayTextColor) || "#fff";
    var disabledTextColor = (style === null || style === void 0 ? void 0 : style.disabledTextColor) || "#ccc";
    var selectedDayBackgroundColor = (style === null || style === void 0 ? void 0 : style.selectedDayBackgroundColor) || "#83bc44";
    var selectedBetweenDayTextColor = (style === null || style === void 0 ? void 0 : style.selectedBetweenDayTextColor) || "#1d1c1d";
    var selectedBetweenDayBackgroundTextColor = (style === null || style === void 0 ? void 0 : style.selectedBetweenDayBackgroundTextColor) || "#F2F2F2";
    var markStyle = {
        width: 30,
        height: 30,
        justifyContent: "center",
        alignItems: "center",
    };
    var betweenStyle = {
        width: "50%",
        height: 30,
        position: "absolute",
        backgroundColor: selectedBetweenDayBackgroundTextColor,
    };
    var dayStyle = {
        color: disabledBeforeToday && isBeforeToday
            ? disabledTextColor
            : isToday
                ? todayColor
                : isHoliday
                    ? holidayColor
                    : dayTextColor,
    };
    switch (type) {
        case "single":
            markStyle = __assign(__assign({}, markStyle), { backgroundColor: selectedDayBackgroundColor, borderRadius: 15 });
            dayStyle = { color: selectedDayTextColor };
            break;
        case "start":
            markStyle = __assign(__assign({}, markStyle), { backgroundColor: selectedDayBackgroundColor, borderRadius: 15 });
            dayStyle = { color: selectedDayTextColor };
            break;
        case "end":
            markStyle = __assign(__assign({}, markStyle), { backgroundColor: selectedDayBackgroundColor, borderRadius: 15 });
            dayStyle = { color: selectedDayTextColor };
            break;
        case "between":
            markStyle = __assign(__assign({}, markStyle), { backgroundColor: selectedBetweenDayBackgroundTextColor, width: "101%" });
            dayStyle = {
                color: isToday
                    ? todayColor
                    : isHoliday
                        ? holidayColor
                        : selectedBetweenDayTextColor,
            };
            break;
        default:
            break;
    }
    return (<>
      {type === "end" ? <View style={[betweenStyle, { left: -1 }]}/> : null}
      {type === "start" ? <View style={[betweenStyle, { right: -1 }]}/> : null}
      {date ? (<View style={markStyle}>
          <Text style={[{ fontSize: 15 }, dayStyle, style === null || style === void 0 ? void 0 : style.dayText]}>
            {moment(date).date()}
          </Text>
        </View>) : null}
      {isToday ? (<Text style={[{ fontSize: 12 }, { color: todayColor }]}>
          {locale.today}
        </Text>) : null}
    </>);
}
function areEqual(prevProps, nextProps) {
    if (prevProps.day.type === nextProps.day.type)
        return true;
    return false;
}
export default memo(Day, areEqual);
