// react-native-calendar-range-picker/dist/Week.js
import React, { memo } from "react";
import { View, StyleSheet, TouchableOpacity } from "react-native";
import Day from "./Day";
function Week(_a) {
    var week = _a.week, locale = _a.locale, handlePress = _a.handlePress, is6Weeks = _a.is6Weeks, disabledBeforeToday = _a.disabledBeforeToday, style = _a.style;
    var renderDayNames = function () {
        var result = [];
        var _loop_1 = function (i) {
            var day = week[i];
            var DayComponent = day.date ? (<TouchableOpacity disabled={disabledBeforeToday && day.isBeforeToday} style={{
                flex: 1,
                height: is6Weeks ? 45 : 50,
                alignItems: "center",
            }} onPress={function () { return handlePress(day.date || ""); }} activeOpacity={1} key={day.date || i}>
          <Day day={day} locale={locale} disabledBeforeToday={disabledBeforeToday} style={style}/>
        </TouchableOpacity>) : (<View style={{ flex: 1, height: is6Weeks ? 45 : 50 }} key={i}/>);
            result.push(DayComponent);
        };
        for (var i = 0; i < 7; i++) {
            _loop_1(i);
        }
        return result;
    };
    return (<View style={[styles.weekContainer, style === null || style === void 0 ? void 0 : style.weekContainer]}>
      {renderDayNames()}
    </View>);
}
function areEqual(prevProps, nextProps) {
    if (JSON.stringify(prevProps.week) === JSON.stringify(nextProps.week))
        return true;
    return false;
}
export default memo(Week, areEqual);
var styles = StyleSheet.create({
    weekContainer: {
        flexDirection: "row",
        alignItems: "center",
    },
});
