// @ui5/webcomponents-react/dist/IllustrationMessageType.js
// Generated file - do not change manually!

/**
 * Different illustration types of Illustrated Message.
 */
var IllustrationMessageType;

(function (IllustrationMessageType) {
  IllustrationMessageType["AddColumn"] = "AddColumn";
  IllustrationMessageType["AddPeople"] = "AddPeople";
  IllustrationMessageType["BalloonSky"] = "BalloonSky";
  IllustrationMessageType["BeforeSearch"] = "BeforeSearch";
  IllustrationMessageType["Connection"] = "Connection";
  IllustrationMessageType["EmptyCalendar"] = "EmptyCalendar";
  IllustrationMessageType["EmptyList"] = "EmptyList";
  IllustrationMessageType["EmptyPlanningCalendar"] = "EmptyPlanningCalendar";
  IllustrationMessageType["ErrorScreen"] = "ErrorScreen";
  IllustrationMessageType["FilterTable"] = "FilterTable";
  IllustrationMessageType["GroupTable"] = "GroupTable";
  IllustrationMessageType["NoActivities"] = "NoActivities";
  IllustrationMessageType["NoData"] = "NoData";
  IllustrationMessageType["NoEntries"] = "NoEntries";
  IllustrationMessageType["NoFilterResults"] = "NoFilterResults";
  IllustrationMessageType["NoMail"] = "NoMail";
  IllustrationMessageType["NoMail_v1"] = "NoMail_v1";
  IllustrationMessageType["NoNotifications"] = "NoNotifications";
  IllustrationMessageType["NoSavedItems"] = "NoSavedItems";
  IllustrationMessageType["NoSavedItems_v1"] = "NoSavedItems_v1";
  IllustrationMessageType["NoSearchResults"] = "NoSearchResults";
  IllustrationMessageType["NoTasks"] = "NoTasks";
  IllustrationMessageType["NoTasks_v1"] = "NoTasks_v1";
  IllustrationMessageType["PageNotFound"] = "PageNotFound";
  IllustrationMessageType["ReloadScreen"] = "ReloadScreen";
  IllustrationMessageType["ResizeColumn"] = "ResizeColumn";
  IllustrationMessageType["SearchEarth"] = "SearchEarth";
  IllustrationMessageType["SearchFolder"] = "SearchFolder";
  IllustrationMessageType["SimpleBalloon"] = "SimpleBalloon";
  IllustrationMessageType["SimpleBell"] = "SimpleBell";
  IllustrationMessageType["SimpleCalendar"] = "SimpleCalendar";
  IllustrationMessageType["SimpleCheckMark"] = "SimpleCheckMark";
  IllustrationMessageType["SimpleConnection"] = "SimpleConnection";
  IllustrationMessageType["SimpleEmptyDoc"] = "SimpleEmptyDoc";
  IllustrationMessageType["SimpleEmptyList"] = "SimpleEmptyList";
  IllustrationMessageType["SimpleError"] = "SimpleError";
  IllustrationMessageType["SimpleMagnifier"] = "SimpleMagnifier";
  IllustrationMessageType["SimpleMail"] = "SimpleMail";
  IllustrationMessageType["SimpleNoSavedItems"] = "SimpleNoSavedItems";
  IllustrationMessageType["SimpleNotFoundMagnifier"] = "SimpleNotFoundMagnifier";
  IllustrationMessageType["SimpleReload"] = "SimpleReload";
  IllustrationMessageType["SimpleTask"] = "SimpleTask";
  IllustrationMessageType["SleepingBell"] = "SleepingBell";
  IllustrationMessageType["SortColumn"] = "SortColumn";
  IllustrationMessageType["SuccessBalloon"] = "SuccessBalloon";
  IllustrationMessageType["SuccessCheckMark"] = "SuccessCheckMark";
  IllustrationMessageType["SuccessHighFive"] = "SuccessHighFive";
  IllustrationMessageType["SuccessScreen"] = "SuccessScreen";
  IllustrationMessageType["Tent"] = "Tent";
  IllustrationMessageType["TntCodePlaceholder"] = "TntCodePlaceholder";
  IllustrationMessageType["TntCompany"] = "TntCompany";
  IllustrationMessageType["TntExternalLink"] = "TntExternalLink";
  IllustrationMessageType["TntFaceID"] = "TntFaceID";
  IllustrationMessageType["TntFingerprint"] = "TntFingerprint";
  IllustrationMessageType["TntLock"] = "TntLock";
  IllustrationMessageType["TntMission"] = "TntMission";
  IllustrationMessageType["TntNoApplications"] = "TntNoApplications";
  IllustrationMessageType["TntNoFlows"] = "TntNoFlows";
  IllustrationMessageType["TntNoUsers"] = "TntNoUsers";
  IllustrationMessageType["TntRadar"] = "TntRadar";
  IllustrationMessageType["TntServices"] = "TntServices";
  IllustrationMessageType["TntSessionExpired"] = "TntSessionExpired";
  IllustrationMessageType["TntSessionExpiring"] = "TntSessionExpiring";
  IllustrationMessageType["TntSuccess"] = "TntSuccess";
  IllustrationMessageType["TntSuccessfulAuth"] = "TntSuccessfulAuth";
  IllustrationMessageType["TntUnlock"] = "TntUnlock";
  IllustrationMessageType["TntUnsuccessfulAuth"] = "TntUnsuccessfulAuth";
  IllustrationMessageType["UnableToLoad"] = "UnableToLoad";
  IllustrationMessageType["UnableToLoadImage"] = "UnableToLoadImage";
  IllustrationMessageType["UnableToUpload"] = "UnableToUpload";
  IllustrationMessageType["UploadCollection"] = "UploadCollection";
})(IllustrationMessageType || (IllustrationMessageType = {}));

export { IllustrationMessageType };
