// @ui5/webcomponents-react/dist/TablePlaceholder.js
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import { ThemingParameters } from '@ui5/webcomponents-react-base/dist/ThemingParameters';
import React, { useMemo } from 'react';
import ContentLoader from 'react-content-loader';

var getArrayOfLength = function getArrayOfLength(len) {
  return Array.from(Array(len).keys());
};

var TableRow = function TableRow(_ref) {
  var columns = _ref.columns,
      y = _ref.y,
      row = _ref.row;
  var columnOffset = 0;
  return /*#__PURE__*/React.createElement(React.Fragment, null, columns.map(function (column, i) {
    var el = /*#__PURE__*/React.createElement("rect", {
      key: "column-".concat(i, "-row-").concat(row),
      x: columnOffset + 2,
      y: y,
      rx: "2",
      ry: "8",
      width: column.totalWidth - 4,
      height: "16"
    });
    columnOffset += column.totalWidth;
    return el;
  }));
};

var TablePlaceholder = function TablePlaceholder(props) {
  var columns = props.columns,
      rows = props.rows,
      style = props.style,
      rowHeight = props.rowHeight,
      tableWidth = props.tableWidth,
      isRtl = props.isRtl;
  var height = rows * rowHeight;
  var innerStyles = useMemo(function () {
    return _objectSpread({
      backgroundColor: ThemingParameters.sapList_Background
    }, style);
  }, [style, ThemingParameters.sapList_Background]);
  return /*#__PURE__*/React.createElement(ContentLoader, {
    style: innerStyles,
    height: height,
    width: tableWidth,
    rtl: isRtl,
    speed: 2,
    backgroundColor: ThemingParameters.sapContent_ImagePlaceholderBackground,
    foregroundColor: ThemingParameters.sapContent_ImagePlaceholderForegroundColor,
    backgroundOpacity: ThemingParameters.sapContent_DisabledOpacity
  }, getArrayOfLength(rows).map(function (_, index) {
    return /*#__PURE__*/React.createElement(TableRow, {
      key: index,
      columns: columns,
      y: rowHeight * index + rowHeight / 2,
      row: index
    });
  }));
};
TablePlaceholder.displayName = 'TablePlaceholder';
TablePlaceholder.defaultProps = {
  rows: 5
};

export { TablePlaceholder };
