// @ui5/webcomponents-react/dist/Text.js
import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import { createUseStyles } from 'react-jss';
import React, { useEffect, forwardRef } from 'react';
import { deprecationNotice } from '@ui5/webcomponents-react-base/dist/Utils';
import { ThemingParameters } from '@ui5/webcomponents-react-base/dist/ThemingParameters';
import clsx from 'clsx';

var useDeprecationNoticeForTooltip = function useDeprecationNoticeForTooltip(component, tooltip) {
  useEffect(function () {
    if (tooltip) {
      deprecationNotice(component, '`tooltip` has been deprecated, please use the native `title` attribute instead.');
    }
  }, [tooltip]);
};

var TextStyles = {
  // Text Style
  text: {
    fontFamily: ThemingParameters.sapFontFamily,
    fontSize: ThemingParameters.sapFontSize,
    fontWeight: 'normal',
    color: ThemingParameters.sapTextColor,
    display: 'inline-block',
    boxSizing: 'border-box',
    whiteSpace: 'pre-line',
    wordWrap: 'break-word',
    maxWidth: '100%'
  },
  renderWhitespace: {
    whiteSpace: 'pre-wrap'
  },
  noWrap: {
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    wordWrap: 'normal',
    overflow: 'hidden',
    '&$renderWhitespace': {
      whiteSpace: 'pre'
    }
  }
};

var _excluded = ["children", "renderWhitespace", "wrapping", "className", "style", "tooltip", "slot"];
var useStyles = createUseStyles(TextStyles, {
  name: 'Text'
});
/**
 * The `Text` component can be used for embedding text into your app. You can hyphenate the text with the use of the `wrapping` prop.
 * <br />__Note:__ Line breaks will always be visualized except when the wrapping property is set to false. In addition, tabs and whitespace can be preserved by setting the renderWhitespace property to true.
 */

var Text = /*#__PURE__*/forwardRef(function (props, ref) {
  var children = props.children,
      renderWhitespace = props.renderWhitespace,
      wrapping = props.wrapping,
      className = props.className,
      style = props.style,
      tooltip = props.tooltip,
      slot = props.slot,
      rest = _objectWithoutProperties(props, _excluded);

  useDeprecationNoticeForTooltip('Text', props.tooltip);
  var classes = useStyles();
  var classNameString = clsx(classes.text, wrapping === false && classes.noWrap, renderWhitespace && classes.renderWhitespace, className);
  return /*#__PURE__*/React.createElement("span", _extends({
    ref: ref,
    style: style,
    className: classNameString,
    title: tooltip,
    slot: slot
  }, rest), children);
});
Text.defaultProps = {
  renderWhitespace: false,
  wrapping: true
};
Text.displayName = 'Text';

export { Text };
