// @ui5/webcomponents-react/wrappers/ObjectPageSubSection.js
import _extends from '@babel/runtime/helpers/extends';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import { ThemingParameters } from '@ui5/webcomponents-react-base/dist/ThemingParameters';
import React, { useEffect, forwardRef } from 'react';
import { createUseStyles } from 'react-jss';
import { deprecationNotice } from '@ui5/webcomponents-react-base/dist/Utils';
import _createClass from '@babel/runtime/helpers/createClass';
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _inherits from '@babel/runtime/helpers/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/getPrototypeOf';
import _wrapNativeSuper from '@babel/runtime/helpers/wrapNativeSuper';
import clsx from 'clsx';

var useDeprecationNoticeForTooltip = function useDeprecationNoticeForTooltip(component, tooltip) {
  useEffect(function () {
    if (tooltip) {
      deprecationNotice(component, '`tooltip` has been deprecated, please use the native `title` attribute instead.');
    }
  }, [tooltip]);
};

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var EmptyIdPropException = /*#__PURE__*/function (_Error) {
  _inherits(EmptyIdPropException, _Error);

  var _super = _createSuper(EmptyIdPropException);

  function EmptyIdPropException(message) {
    var _this;

    _classCallCheck(this, EmptyIdPropException);

    _this = _super.call(this, message);
    _this.name = _this.constructor.name;
    return _this;
  }

  return _createClass(EmptyIdPropException);
}( /*#__PURE__*/_wrapNativeSuper(Error));

var _excluded = ["children", "id", "titleText", "className", "style", "tooltip"];
var styles = {
  objectPageSubSection: {
    '&:focus': {
      outline: "".concat(ThemingParameters.sapContent_FocusWidth, " ").concat(ThemingParameters.sapContent_FocusStyle, " ").concat(ThemingParameters.sapContent_FocusColor),
      outlineOffset: "calc(-1 * ".concat(ThemingParameters.sapContent_FocusWidth, ")")
    }
  },
  objectPageSubSectionHeaderTitle: {
    fontSize: ThemingParameters.sapFontHeader5Size,
    color: ThemingParameters.sapGroup_TitleTextColor,
    marginBottom: '0.5rem'
  },
  subSectionContent: {
    padding: '1rem 0 3rem 0'
  }
};
var useStyles = createUseStyles(styles, {
  name: 'ObjectPageSubSection'
});
/**
 * Second-level information container of an `ObjectPage`.<br />
 * __Note:__ This component should only be used inside an `ObjectPageSection` component.
 */

var ObjectPageSubSection = /*#__PURE__*/forwardRef(function (props, ref) {
  var children = props.children,
      id = props.id,
      titleText = props.titleText,
      className = props.className,
      style = props.style,
      tooltip = props.tooltip,
      rest = _objectWithoutProperties(props, _excluded);

  useDeprecationNoticeForTooltip('ObjectPageSubSection', props.tooltip);

  if (!id) {
    throw new EmptyIdPropException('ObjectPageSubSection requires a unique ID property!');
  }

  var htmlId = "ObjectPageSubSection-".concat(id);
  var classes = useStyles();
  var subSectionClassName = clsx(classes.objectPageSubSection, className);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    role: "region",
    style: style,
    title: tooltip,
    tabIndex: -1
  }, rest, {
    className: subSectionClassName,
    id: htmlId
  }), /*#__PURE__*/React.createElement("div", {
    role: "heading",
    "aria-level": 4,
    className: classes.objectPageSubSectionHeaderTitle,
    "data-component-name": "ObjectPageSubSectionTitleText"
  }, titleText), /*#__PURE__*/React.createElement("div", {
    className: classes.subSectionContent,
    "data-component-name": "ObjectPageSubSectionContent"
  }, children));
});
ObjectPageSubSection.displayName = 'ObjectPageSubSection';

export { ObjectPageSubSection };
