// @ui5/webcomponents-react/wrappers/ThemeProvider.js
import { cssVariablesStyles } from '@ui5/webcomponents-react-base/dist/CssSizeVariables';
import { ThemingParameters } from '@ui5/webcomponents-react-base/dist/ThemingParameters';
import React, { useEffect } from 'react';
import { createUseStyles, ThemeProvider as ThemeProvider$1 } from 'react-jss';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { GlobalStyleClasses } from '@ui5/webcomponents-react/dist/GlobalStyleClasses';

var GlobalStyleClassesStyles = {
  '@global': _defineProperty({}, ".".concat(GlobalStyleClasses.sapScrollBar, ":not([data-native-scrollbar])"), {
    '&::-webkit-scrollbar': {
      backgroundColor: ThemingParameters.sapScrollBar_TrackColor,
      '&:horizontal': {
        height: ThemingParameters.sapScrollBar_Dimension
      },
      '&:vertical': {
        width: ThemingParameters.sapScrollBar_Dimension
      }
    },
    '&::-webkit-scrollbar-thumb': {
      backgroundColor: ThemingParameters.sapScrollBar_FaceColor,
      borderRadius: '0.25rem',
      '&:hover': {
        backgroundColor: ThemingParameters.sapScrollBar_Hover_FaceColor
      },
      '&:horizontal': {
        height: ThemingParameters.sapScrollBar_Dimension
      },
      '&:vertical': {
        width: ThemingParameters.sapScrollBar_Dimension
      }
    },
    '&::-webkit-scrollbar-corner': {
      backgroundColor: ThemingParameters.sapScrollBar_TrackColor
    }
  })
};

var useStyles = createUseStyles(GlobalStyleClassesStyles);

var cssVarsPonyfillNeeded = function cssVarsPonyfillNeeded() {
  return !!window.CSSVarsPonyfill;
};

// inject the size variables first before the ThemeProvider Component is mounted, otherwise there will be some flickering
if (!document.querySelector('style[data-ui5-webcomponents-react-sizes]')) {
  var variables = document.createElement('style');
  variables.setAttribute('data-ui5-webcomponents-react-sizes', '');
  variables.innerHTML = cssVariablesStyles;
  document.head.appendChild(variables);
}

var ThemeProvider = function ThemeProvider(props) {
  var children = props.children;
  useStyles();
  useEffect(function () {
    if (cssVarsPonyfillNeeded()) {
      window.CSSVarsPonyfill.cssVars({
        rootElement: document.head,
        include: 'style[data-ui5-webcomponents-react-sizes],style[data-jss]',
        watch: true,
        silent: true
      });
    }
  }, []);
  return /*#__PURE__*/React.createElement(ThemeProvider$1, {
    theme: ThemingParameters
  }, children);
};

ThemeProvider.displayName = 'ThemeProvider';

export { ThemeProvider };
