// @ui5/webcomponents-react-base/dist/spacing.js
import _objectSpread from '@babel/runtime/helpers/objectSpread2';

var TINY_SPACE = '0.5rem';
var SMALL_SPACE = '1rem';
var MEDIUM_SPACE = '2rem';
var LARGE_SPACE = '3rem';
/*
 * ###################################################
 * Margin Classes
 * ###################################################
 */
// All Around Margins

var sapUiTinyMargin = {
  margin: TINY_SPACE
};
var sapUiSmallMargin = {
  margin: SMALL_SPACE
};
var sapUiMediumMargin = {
  margin: MEDIUM_SPACE
};
var sapUiLargeMargin = {
  margin: LARGE_SPACE
}; // Single Side Margins
// Tiny

var sapUiTinyMarginTop = {
  marginTop: TINY_SPACE
};
var sapUiTinyMarginBottom = {
  marginBottom: TINY_SPACE
};
var sapUiTinyMarginBegin = {
  marginLeft: TINY_SPACE
};
var sapUiTinyMarginEnd = {
  marginRight: TINY_SPACE
}; // Small

var sapUiSmallMarginTop = {
  marginTop: SMALL_SPACE
};
var sapUiSmallMarginBottom = {
  marginBottom: SMALL_SPACE
};
var sapUiSmallMarginBegin = {
  marginLeft: SMALL_SPACE
};
var sapUiSmallMarginEnd = {
  marginRight: SMALL_SPACE
}; // Medium

var sapUiMediumMarginTop = {
  marginTop: MEDIUM_SPACE
};
var sapUiMediumMarginBottom = {
  marginBottom: MEDIUM_SPACE
};
var sapUiMediumMarginBegin = {
  marginLeft: MEDIUM_SPACE
};
var sapUiMediumMarginEnd = {
  marginRight: MEDIUM_SPACE
}; // Large

var sapUiLargeMarginTop = {
  marginTop: LARGE_SPACE
};
var sapUiLargeMarginBottom = {
  marginBottom: LARGE_SPACE
};
var sapUiLargeMarginBegin = {
  marginLeft: LARGE_SPACE
};
var sapUiLargeMarginEnd = {
  marginRight: LARGE_SPACE
}; // Two Sided Margin
// Tiny

var sapUiTinyMarginBeginEnd = _objectSpread(_objectSpread({}, sapUiTinyMarginBegin), sapUiTinyMarginEnd);
var sapUiTinyMarginTopBottom = _objectSpread(_objectSpread({}, sapUiTinyMarginTop), sapUiTinyMarginBottom); // Small

var sapUiSmallMarginBeginEnd = _objectSpread(_objectSpread({}, sapUiSmallMarginBegin), sapUiSmallMarginEnd);
var sapUiSmallMarginTopBottom = _objectSpread(_objectSpread({}, sapUiSmallMarginTop), sapUiSmallMarginBottom); // Medium

var sapUiMediumMarginBeginEnd = _objectSpread(_objectSpread({}, sapUiMediumMarginBegin), sapUiMediumMarginEnd);
var sapUiMediumMarginTopBottom = _objectSpread(_objectSpread({}, sapUiMediumMarginTop), sapUiMediumMarginBottom); // Large

var sapUiLargeMarginBeginEnd = _objectSpread(_objectSpread({}, sapUiLargeMarginBegin), sapUiLargeMarginEnd);
var sapUiLargeMarginTopBottom = _objectSpread(_objectSpread({}, sapUiLargeMarginTop), sapUiLargeMarginBottom); // Remove Margins

var sapUiNoMargin = {
  margin: '0 !important'
}; // Remove Single Side Margins

var sapUiNoMarginTop = {
  marginTop: '0 !important'
};
var sapUiNoMarginBottom = {
  marginBottom: '0 !important'
};
var sapUiNoMarginBegin = {
  marginBegin: '0 !important'
};
var sapUiNoMarginEnd = {
  marginEnd: '0 !important'
}; // Force Auto Width

var sapUiForceWidthAuto = {
  width: 'auto !important'
};
var sapUiResponsiveMargin = {
  '@media(max-width:599px)': {
    margin: '0 0 1rem 0'
  },
  '@media (min-width:600px) and (max-width:1023px)': {
    margin: '1rem !important'
  },
  '@media (min-width:1024px) and (max-width: 1439px)': {
    margin: '1rem 2rem !important'
  },
  '@media (min-width:1440px)': {
    margin: '1rem 3rem !important'
  }
};
var sapUiTinyNegativeMarginBeginEnd = {
  margin: '0 -0.5rem !important'
};
var sapUiSmallNegativeMarginBeginEnd = {
  margin: '0 -1rem !important'
};
var sapUiMediumNegativeMarginBeginEnd = {
  margin: '0 -2rem !important'
};
var sapUiLargeNegativeMarginBeginEnd = {
  margin: '0 -3rem !important'
};
/*
 * ###################################################
 * Padding Classes
 * ###################################################
 */

var sapUiNoContentPadding = {
  padding: '0 !important'
};
var sapUiContentPadding = {
  padding: '1rem'
}; // Two Sided Padding

var sapUiTinyPaddingBeginEnd = {
  paddingLeft: '0.5rem !important',
  paddingRight: '0.5rem !important'
};
var sapUiSmallPaddingBeginEnd = {
  paddingLeft: '1rem !important',
  paddingRight: '1rem !important'
};
var sapUiMediumPaddingBeginEnd = {
  paddingLeft: '2rem !important',
  paddingRight: '2rem !important'
};
var sapUiLargePaddingBeginEnd = {
  paddingLeft: '3rem !important',
  paddingRight: '3rem !important'
};
var sapUiResponsiveContentPadding = {
  '@media(max-width:599px)': {
    paddingLeft: '1rem',
    paddingRight: '1rem'
  },
  '@media (min-width:600px) and (max-width:1023px)': {
    paddingLeft: '2rem',
    paddingRight: '2rem'
  },
  '@media (min-width:1024px) and (max-width: 1439px)': {
    paddingLeft: '2rem',
    paddingRight: '2rem'
  },
  '@media (min-width:1440px)': {
    paddingLeft: '3rem',
    paddingRight: '3rem'
  }
};

export { sapUiContentPadding, sapUiForceWidthAuto, sapUiLargeMargin, sapUiLargeMarginBegin, sapUiLargeMarginBeginEnd, sapUiLargeMarginBottom, sapUiLargeMarginEnd, sapUiLargeMarginTop, sapUiLargeMarginTopBottom, sapUiLargeNegativeMarginBeginEnd, sapUiLargePaddingBeginEnd, sapUiMediumMargin, sapUiMediumMarginBegin, sapUiMediumMarginBeginEnd, sapUiMediumMarginBottom, sapUiMediumMarginEnd, sapUiMediumMarginTop, sapUiMediumMarginTopBottom, sapUiMediumNegativeMarginBeginEnd, sapUiMediumPaddingBeginEnd, sapUiNoContentPadding, sapUiNoMargin, sapUiNoMarginBegin, sapUiNoMarginBottom, sapUiNoMarginEnd, sapUiNoMarginTop, sapUiResponsiveContentPadding, sapUiResponsiveMargin, sapUiSmallMargin, sapUiSmallMarginBegin, sapUiSmallMarginBeginEnd, sapUiSmallMarginBottom, sapUiSmallMarginEnd, sapUiSmallMarginTop, sapUiSmallMarginTopBottom, sapUiSmallNegativeMarginBeginEnd, sapUiSmallPaddingBeginEnd, sapUiTinyMargin, sapUiTinyMarginBegin, sapUiTinyMarginBeginEnd, sapUiTinyMarginBottom, sapUiTinyMarginEnd, sapUiTinyMarginTop, sapUiTinyMarginTopBottom, sapUiTinyNegativeMarginBeginEnd, sapUiTinyPaddingBeginEnd };
