// @ui5/webcomponents-react-base/dist/StyleClassHelper.js
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _createClass from '@babel/runtime/helpers/createClass';
import _defineProperty from '@babel/runtime/helpers/defineProperty';

/**
 * Created by d059190 at 16.03.18
 */
var StyleClassHelper = /*#__PURE__*/function () {
  function StyleClassHelper() {
    _classCallCheck(this, StyleClassHelper);

    _defineProperty(this, "classes", []);

    console.warn("StyleClassHelper is deprecated! Please use clsx instead.");

    for (var _len = arguments.length, classes = new Array(_len), _key = 0; _key < _len; _key++) {
      classes[_key] = arguments[_key];
    }

    this.classes = [].concat(classes);
  }

  _createClass(StyleClassHelper, [{
    key: "put",
    value: function put() {
      var _this$classes;

      (_this$classes = this.classes).push.apply(_this$classes, arguments);

      return this;
    }
  }, {
    key: "putIfPresent",
    value: function putIfPresent() {
      for (var _len2 = arguments.length, clazzes = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        clazzes[_key2] = arguments[_key2];
      }

      for (var _i = 0, _clazzes = clazzes; _i < _clazzes.length; _i++) {
        var clazz = _clazzes[_i];

        if (clazz) {
          this.classes.push(clazz);
        }
      }

      return this;
    }
  }, {
    key: "valueOf",
    value: function valueOf() {
      return this.className;
    }
  }, {
    key: "toString",
    value: function toString() {
      return this.className;
    }
  }, {
    key: "className",
    get: function get() {
      return this.classes.join(' ');
    }
  }], [{
    key: "of",
    value: function of() {
      var _StyleClassHelper;

      return (_StyleClassHelper = new StyleClassHelper()).put.apply(_StyleClassHelper, arguments);
    }
  }]);

  return StyleClassHelper;
}();

export { StyleClassHelper };
